### DISCUSSION TITLE
Panopticon on Synology

### DISCUSSION THREAD

**Discussion by nikosdion** [akeeba/panopticon#466] [Show and tell]
# Panopticon on Synology

I've written in the documentation that Panopticon is very lightweight and can be installed pretty much anywhere. Between that and my willingness to eat my own dogfood, I decided to deploy my own site monitoring on my ageing Synology DS218+ which has a feeble, 7-year-old [Celeron J3355](https://ark.intel.com/content/www/us/en/ark/products/95597/intel-celeron-processor-j3355-2m-cache-up-to-2-5-ghz.html) and all of 6GiB or RAM, populated entirely with sluggish Western Digital Red HDDs (yep, spinning rust, no SSD for this old fellow).

I did. It's decently fast. More than that, it's _perfectly usable_ and I am currently using it to monitor my sites. I no longer need to pay for a separate server.

## What Do We Need?

![](https://media.giphy.com/media/OsItQTbcxuIyQ/giphy.gif)

Deploying a web application to the Synology NAS means that we need a database server, a web server, a decently recent version of PHP, a way to contact our server from the outside, and the HTTPS certificates to go with it (because passkeys only work under HTTPS).

I decided against using Docker since Synology won't set up rootless Docker when you install Container Manager, and I did not want to have to be root to set this all up. This means using Web Station for the web server, along with the Synology-provided PHP 8.2 and MariaDB 10.11 packages.

Accessing the NAS from the outside world was already possible using dynamic DNS handled through the Synology DDNS service, and HTTPS certificates are issued by Let's Encrypt and handled automatically.

Before you venture forward, let me tell you about the three snags I hit along the way.

Snag 1: CRON setup. Synology Disk Station Manager (a.k.a. DSM) – the Linux-based Operating System of every Synology NAS – does have CRON jobs, but they are abstracted behind the weird contraption called the Task Scheduler. While it does create CRON jobs, they trigger a secondary application (the task scheduler) which does _additional scheduling_ on its own. The result is that tasks scheduled through the user interface only activate 24 hours after they are created or changed in the user interface _at the earliest_. Therefore, I had to get my hands dirty and use some command line magic to work around this limitation.

Snag 2: HTTPS (TLS Certificate) setup. By default, Synology DSM will use Let's Encrypt to issue a TLS certificate which is only valid for the main public domain you use with your NAS. We need to use Synology's DDNS service to get a subdomain, have our site monitoring set up in a secondary subdomain of that, and tell Synology's DSM to issue a TLS certificate via Let's Encrypt which applies to all secondary subdomains of the subdomain linked with Synology's DDNS service. This only works with Synology's DDNS service, though. If you have your own domain you'll have to provision your own TLS certificates yourself.

Snag 3: Email setup. I thought I'd use the PHP Mail option, which relays email through the mail notification address set up in DSM. As I discovered, this is broken if you use Gmail or Hotmail / Microsoft Office 365 – broken at the Synology DSM level, not something I can fix. Moreover, DSM does not provide a `sendmail` command by default, meaning that option was out of the window as well. Therefore, I had to use SMTP.

## Installing Panopticon On The Synology NAS

![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExbmc2ZXd4MHM4amR2Z2Nta2Q4ZGo5MGVnbTF6ZTNvZXk1dHBtY2U0NCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/gZuxOq7zSL5DO/giphy.gif)

Without further ado, let's get down to business. I will guide you through the logical sequence of steps needed.

### Dynamic DNS

To access our site monitoring from the Internet, e.g. from a laptop, tablet, or phone while not in the office, we need _something_ to link a domain name to the (often changing) IP address of our router. This is where Dynamic DNS comes into play. If you have already set up DDNS _with the Synology DDNS service_ you can skip this step, noting that the domain name I chose for this example is `example.myDS.me`.

Go to Control Panel, External Access, DDNS. Click on Add.

Under Service Provider choose Synology.

In the hostname, there are two fields. To get the domain `example.myDS.me` I described above one would enter `example` in the first text field, and then choose `myDS.me` from the dropdown next to it. 

Leave the other options as is and save.

### Find Your Port

In most cases, ISPs frown upon their clients exposing port 80 (HTTP) and 443 (HTTPS) on the domestic or small business connections they provide. They don't want you to run your own servers. To circumvent that, DSM uses non-privileged public ports (port numbers above 1024) for these services, typically 5000 for HTTP and 5001 for HTTPS. We need to know which are these ports to construct our Panopticon login URL later.

Go to Control Panel, External Access, Router Configuration.

Here you can see which Router Port is assigned to the Management UI service. Note it down.

### Set Up HTTPS (TLS Certificates)

Synology DSM can automatically issue and renew HTTPS certificates for us at no cost, using Let's Encrypt. By default, however, it only issues a certificate for the main subdomain you access your Synology NAS with, e.g. `example.myDS.me`. Panopticon will run at a second level subdomain, e.g. `monitoring.example.myDS.me`. We'll need to tell Synology to request and use a certificate which will be valid for all of our second-level subdomains.

To do that, go to Control Panel, Security, Certificate. Click on Add.

Select “Add a new certificate”. Click on Next.

Enter any description you want, e.g. “Main HTTPS Certificate”. Select “Get a certificate from Let's Encrypt”. Check the “Set as default certificate” box. Click on Next.

In the “Domain name” area enter your Synology DDNS domain name, in our example it's `example.myDS.me`. Enter your email address under “Email”. In the Subject Alternative Name you need to enter your Synology DDNS domain name preceded by `*.`, e.g. `*.example.myDS.me`. _This is only possible if you are using Synology DDNS_. Click on Done.

### Install Packages

Go to Package Center. Look for and install the following packages:

* PHP 8.2
* Apache HTTP Server 2.4
* Web Station
* MariaDB
* phpMyAdmin

The latter is optional, but it will make your life managing database on your Synology NAS incredibly easier.

_ℹ️ Note_: Yes, I told you to look for PHP 8.2, even though the latest version of PHP is 8.3. Unfortunately, Synology DSM only offers slightly dated versions of web server software, especially PHP – which is weird, since most of their server-side software is written in PHP. Oh, well. PHP 8.2 isn't terrible and will be good for a couple of years after the time of this writing, so at least there's that.

### Set Up Your Database

We'll need to create a database for Panopticon. This is a two-step process.

![](https://media.giphy.com/media/Kzh5CXQ91XaQSVSUBc/giphy.gif)

#### Set Up A MariaDB Root Password

From the app selector, go to MariaDB.

There should be a button to set or reset the root password. Use it to create a password for your database's `root` user.

_🔐 Security Note:_ I strongly recommend creating a random password, e.g. with [Random.org](https://www.random.org/passwords/?num=1&len=32&format=html&rnd=new), and store it in a password manager.

#### Create A Database And User

From the app selector, go to phpMyAdmin.

ℹ️ If you get an error about connecting to the server… **Don't Panic!** The problem is that either ports 80 and 443 are not forwarded, or your ISP is blocking ports 80 and 443 which are required to access this application. If the Synology DSM is on your network change your URL from `http://example.myDS.me/synology` to`http://10.0.0.1/synology` where `10.0.0.1` is the local network IP address of your NAS. Where to find it? Control Panel, Info Center, Network; look under LAN, IP address.

Enter Username `root` and Password what you set up in the step above, then click on Log In.

Click “User Accounts” at the top, then click on the “Add user account” link.

In the new page, enter the following:

* User name: panopticon
* Host Name: From the drop-down select “Local”
* Password: Create a random password and note it down.
* Authentication plugin: Native MySQL authentication
* Create database with same name and grant all privileges: check
* Grant all privileges on wildcard name (username\_%): check

Click on the “Go” button.

This creates both a database user, and a database, to which our user is added with the appropriate permissions.

### Create A Web Application

We will now create the Web Station application which will be serving Panopticon on our NAS.

This is a multistep process which may appear daunting at first. That's why I wrote up these instructions. To give you an idea of what we're doing, we're creating three things:

* Script Language Settings tell Synology DSM how to configure PHP for our Panopticon installation.
* The Web Service tell Synology DSM where Panopticon's files will be installed, and which Script Language Settings to use to run them.
* The Web Portal tells Synology DSM the domain name and port which we'll be using to access Panopticon.

One builds upon the other, and that's the reason we have to create them in an order that looks “upside-down” when you look at Synology's Web Station interface. The interface is geared towards management of what you've already set up, not an intuitive setup process.

#### Create Script Language Settings

From the app selector, go to Web Station.

Click on the “Script Language Settings” tab.

In that tab, click on the “Create” button.

In the first page, enter the following:
* Profile Name: PHP Panopticon
* Description: My PHP settings for Panopticon
* PHP version: PHP 8.2
* Enable PHP cache: check

Click on the Next button. In the new page, check the boxes next to the following items:

* bcmath
* bz2
* curl
* ftp
* gd
* gettext
* gmp
* imap
* intl
* mysqli
* openssl
* pdo_mysql
* phar
* sodium
* ssh2
* xsl
* zip
* zlib

These are more PHP extensions than what you currently need. I've included some extensions which might be needed in the future – basically, all extensions I have ever used in software I've written.

Click on the Next button.

Leave the FPM settings as-is on this page. Just click on the Next button.

In this new page change the following settings:

* `upload_max_filesize`: 80M
* `post_max_size`: 80M

Technically, these changes are not currently necessary, but will be needed when some planned features are implemented. Click on the Next button.

Click on the Create button. Phew.

#### Create A Web Service

From the app selector, go to Web Station.

Click on the “Web Service” tab. On that tab, click on the “Create” button.

Select “Native script language website”.

In the left-hand Service drop-down select PHP 8.2. In the right-hand Service drop-down select “PHP Panopticon” – yes, the one we created in the previous, long step.

Click on the Next button.

Enter the following:
* Name: `panopticon` (MUST be all lowercase, no spaces)
* Description: Site monitoring with Akeeba Panopticon
* Document Root: click on Browse, and nvaigate to the web folder. Click on Create Folder and enter the fodler name `panopticon`, then click on OK. Select the newly created `panopticon` folder and click on Select.
* HTTP back-end server: Apache HTTP Server 2.4

Click on the Next button.

Click on the Create button.

#### Create A Customised Web Portal

Web Station

WEb Portal

Create

Select “Web service portal”

* Service: Panopticon
* Portal type: Name-based
* Hostname: `monitoring.example.myDS.me`
* Port:
  * Uncheck `80 / 443`
  * Check HTTPS and enter the port we found out earlier, e.g. `5001`
* HTTPS settings:
  * Check the “HSTS” box

Click on the Create button

Remember, the URL for Panopticon will be `https://` followed by the Hostname, followed by a colon, followed by the HTTP port. In our example it's `https://monitoring.example.myDS.me:5001`.

#### Make Sure The Correct HTTP Certificate Is Used

Go back to Control Panel, Security, Certificate.

Click on Settings. 

Scroll down and find your newly created Web Portal domain name, e.g. `monitoring.example.myDS.me:5001`. Make sure the drop-down value next to it is the main certificate we created earlier. You might have to click on the drop-down to inspect its information in more detail.

If the wrong certificate is selected, select the correct one, then click on OK.

### Install Panopticon

First, [download Panopticon](https://github.com/akeeba/panopticon/releases/latest) from GitHub.

Open your Synology File Station and go to the `web` folder, then double-click on the `panopticon` folder.

Upload the Panopticon ZIP file into this folder. Then, right-click on it and select Extract, Extract Here. After the extraction is over you can remove the ZIP file.

The installation process is as described in the [Panopticon wiki](https://github.com/akeeba/panopticon/wiki/Install-Panopticon).

When you get to the database page, remember that your connection information is as follows:
* Database type: MySQLi
* Database server host name: `localhost`
* User name: `panopticon`
* Password: the password you chose when creating the database in phpMyAdmin
* Database name: `panopticon`

Leave the other database options as-is.

Proceed with the installation until you reach the “Set up a CRON job” page.

### Set Up CRON jobs

Panopticon requires one or more CRON jobs to periodically execute its various automation actions. 

One way to create them is using Synology's Task Scheduler. _I do not recommend this method_ as explained in Snag 1. To make it very clear, yes, it is easier to use, but any task you create _or modify_ will only start running after 24 hours. This means that when you try to set up Panopticon, or try to change its CRON jobs, you have to wait a full day at least to see any change. That's… insane.

So, this is where it's getting a bit tough. We'll have to use the command line to edit the crontab file. To make matters even more complicated, Synology does not provide a command-line text editor. Therefore, we will need to install a third party package to do this. Hold onto your hats, it's gonna be a wild ride!

![](https://media.giphy.com/media/3knKct3fGqxhK/giphy.gif)

#### Set Up A Third Party Repository

Go to Package Center and click on Settings.

Click on the Package Sources tab.

Click on Add and enter the following information:
* Name: `SynoCommunity`
* Location: `http://packages.synocommunity.com/`

Click on the OK button.

Optionally, you can add yet another source by clicking the Add button and entering the following information:
* Name: `cpHub`
* Location: `https://www.cphub.net`

#### Install A Text Editor

Go to the Package Center and click on the Community tab.

Search for the package `SynoCli File Tools` and install it. This, among other useful things, includes the text editor [`nano`](https://www.nano-editor.org).

#### Enabling SSH

If you have not done already, you will need to enable SSH on your NAS.

Go to Control Panel, Terminal & SNMP.

Click on the Terminal tab and check the “Enable SSH Service” checbox. In the Port field enter 2323.

Click on the “Advanced Settings” button and select the High option, then click on Save. Now back in the Control Panel window, click on Apply.

#### Connect To Your DNS Via SSH

For the next step you will need an SSH client. My recommendations are as follows:
* Windows: [PuTTY](https://putty.org).
* macOS: PuTTY ([via HomeBrew](https://formulae.brew.sh/formula/putty#default)), [Termius](https://apps.apple.com/us/app/termius-ssh-sftp-client/id1176074088?mt=12), or the `ssh` command which is already preinstalled and available in the Terminal app.
* Linux: the `ssh` command, typically part of `openssh` package.

Either way, you can connect to your NAS' SSH server using the hostname you normally access your NAS with (e.g. `example.myDS.me`), port `2323`, the username you log into your NAS (e.g. `myuser`), and the password you log into your NAS. With these example connection details when using SSH in the command line you would connect to your NAS with `ssh myuser@example.myDS.com -p 2323`, then enter your password when prompted.

#### Setting up the CRON job

Once connected to your NAS over SSH (feel free to enunciate “I'm in” in a husky voice) it's time to _finally_ get down to business.

Run the following command:

```bash
sudo nano /etc/crontab
```

Enter your user's password when propmpted.

You will now see a pretty long file. Add a new line at the bottom which looks like this:

```cronexp
*   *   *   *   *   root    /bin/su -c "/usr/local/bin/php82 /volume1/web/panopticon/cli/panopticon.php task:run --loop" myuser >/dev/null 2>&1
```

Here's what you need to change:
* Replace `/volume1/web/panopticon/cli/panopticon.php` with the path shown to you in Panopticon's installation page on your browser
* Replace `myuser` with your username

Press CTRL-X, then Y to exit, saving your changes to the file.

Them run the command:

```bash
sudo systemctl restart crond
```

This tells Synology DSM to reload the CRON job definitions from this file, and start executing the CRON jobs in it.

After a minute or so you will see the Panopticon installation page proceed with the installation.

![](https://media.giphy.com/media/Na33dsU2umStO/giphy.gif)

### Set Up Email In Panopticon

As per Snag 3 I explained at the very beginning, setting up email in Panopticon running under Synology is _tricky_. If you get it wrong you'll end up with errors showing up in automated tasks, you will not receive emails, and you may experience occasional hangs (503 errors) when trying to use Panopticon.

Remember, you can set up email in Panopticon's Administration, System Configuration page, under the Email tab. Always **Save** your changes _before_ clicking the Send Test Email Message button.

#### Using The “PHP Mail” option

When you try to use Panopticon's “PHP Mail” option to send emails, things may or may not work.

By default, PHP is configured to use `ssmtp` to relay email messages sent with PHP's `mail()` function. This is supposed to send the emails with the same email configuration you've set up in Control Panel, Notification, Email in the DSM.

If you have set up email notifications with a regular SMTP server it will work just fine.

However, if you tried using Gmail, or Outlook – both using the XOAUTH authentication protocol – you will receive no email. The thing is that `ssmtp` is obsolete software which predates the XOAUTH protocol and does not support it. Basically, the software Synology provides for PHP to talk to mail servers does not speak the same language as Gmail's and Hotmail's / Microsoft Office 365's mail servers.

If you want to use the PHP Mail function you will need to set up the notifications email to use a Custom SMTP Server, meaning the SMTP server your hosting provider set up for you on your site. Sorry. Life sucks.

One more weird snag with this solution is that the `ssmtp` package used by Synology _only supports 7-bit US ASCII characters_. If you try to send an email with non-ASCII characters – so, basically, in any language other than English – you end up with corrupt characters. Considering that my language uses a completely different alphabet, let's just say that the end result was hilariously wrong and I gave up on using this method before I could finish laughing at the absurdity of it all.

#### Using The Sendmail option

On most Synology NAS installations this will NOT work at all.

![](https://media.giphy.com/media/Yycc82XEuWDaLLi2GV/giphy.gif)

Synology DSM does not provide a `sendmail` command _unless_ you have installed the Synology MailPlus Server package. However, the function of that package is to make the Synology NAS operate as a fully fledged mailserver. If you have already set it up, sure, go ahead and use Sendmail. If you have not done so already, **don't**.

#### Using The SMTP Option

For whatever reason, I could not get the standard TCP/IP port 25 (unencrypted SMTP) to work with the Synology NAS, even though I could connect to the same server from another Panopticon installation on my dev machine, on the same Internet connection, just fine. Same deal with port 465, but that _may_ have been a problem with PHPMail – the mail library used in Panopticon – only supporting STARTTLS and not straight up TLS (SMTPS) connections correctly, or at the very least not agreeing with the TLS implementation of any mail server I had access to.

In any case, my solution was rather straightforward. I used port 587 with my SMTP server which implicitly tells PHPMail to use the STARTTLS protocol to talk to the SMTP server. Since my SMTP server (Amazon SES) supports that method, it all worked fine. 

So, if your SMTP email set up doesn't work, try port 587 to force the use of STARTTLS, assuming your mail server supports it.

## Is It Worth The Effort?

If you have a Synology NAS – which is very likely in some parts of Europe (Hallo, meine deutschen Freunde! Wie geht's?) – it's absolutely worth the effort. You'll spend an afternoon. In return, you can monitor all your sites using a self-hosted site monitoring server which costs you nothing, maintaining ownership of your data.

If you're apprehensive about having to use the command line to set up the CRON jobs, don't be. You'll just be using two canned commands and a text editor. You're building sites, a far more complicated task. Just believe in yourself.

![Ted Lasso pointing to the BELIEVE sign above the coaches' office door in a short GIF from the eponymous TV show](https://media.giphy.com/media/DEZA7FlHbMesUF1jm9/giphy.gif "Believe in Believe")

If you don't have a Synology NAS, well, the takeaway is that Panopticon will run great and be perfectly usable even on hardware verging on obsolescence. If you have some old hardware lying around – even if it has a slow CPU, not a lot of memory, and spinning hard drives – you can totally repurpose it as a site monitoring server.

**Comment by brianteeman** [comment]
nice - when I get a free couple of hourss I was planning on trying my pi-zero
